File: libs/moqtail-ts/src/model/control/subscribe.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Location } from '../common/location'
import { Tuple } from '../common/tuple'
import { KeyValuePair } from '../common/pair'
import { ControlMessageType, FilterType, GroupOrder } from '../control/constant'
import { FullTrackName } from '../data'
import { MessageParameter } from '../parameter/message_parameter'
import { SubscriptionFilter } from '../parameter/message/subscription_filter'
import { Forward } from '../parameter/message/forward'
import { GroupOrderParam } from '../parameter/message/group_order_param'

export class Subscribe {
  private constructor(
    public requestId: bigint,
    public fullTrackName: FullTrackName,
    public parameters: MessageParameter[],
  ) {}

  static newNextGroupStart(requestId: bigint, fullTrackName: FullTrackName, parameters: MessageParameter[]): Subscribe {
    return new Subscribe(requestId, fullTrackName, [...parameters, new SubscriptionFilter(FilterType.NextGroupStart)])
  }

  static newLatestObject(requestId: bigint, fullTrackName: FullTrackName, parameters: MessageParameter[]): Subscribe {
    return new Subscribe(requestId, fullTrackName, [...parameters, new SubscriptionFilter(FilterType.LatestObject)])
  }

  static newAbsoluteStart(
    requestId: bigint,
    fullTrackName: FullTrackName,
    startLocation: Location,
    parameters: MessageParameter[],
  ): Subscribe {
    return new Subscribe(requestId, fullTrackName, [
      ...parameters,
      new SubscriptionFilter(FilterType.AbsoluteStart, startLocation),
    ])
  }

  static newAbsoluteRange(
    requestId: bigint,
    fullTrackName: FullTrackName,
    startLocation: Location,
    endGroup: bigint,
    parameters: MessageParameter[],
  ): Subscribe {
    if (endGroup < startLocation.group) {
      throw new Error('End Group must be >= Start Group')
    }
    return new Subscribe(requestId, fullTrackName, [
      ...parameters,
      new SubscriptionFilter(FilterType.AbsoluteRange, startLocation, endGroup),
    ])
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.Subscribe)

    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putBytes(this.fullTrackName.serialize().toUint8Array())

    payload.putVI(this.parameters.length)
    for (const param of this.parameters) {
      payload.putBytes(param.toKeyValuePair().serialize().toUint8Array())
    }

    const payloadBytes = payload.toUint8Array()
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)

    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): Subscribe {
    const requestId = buf.getVI()
    const fullTrackName = buf.getFullTrackName()

    const paramCount = Number(buf.getVI())
    const parameters: MessageParameter[] = []
    for (let i = 0; i < paramCount; i++) {
      const kvp = KeyValuePair.deserialize(buf)
      const param = MessageParameter.fromKeyValuePair(kvp)
      if (param !== undefined) parameters.push(param)
    }

    return new Subscribe(requestId, fullTrackName, parameters)
  }
}

if (import.meta.vitest) {
  const { describe, it, expect } = import.meta.vitest

  function buildTestSubscribe(): Subscribe {
    return Subscribe.newAbsoluteRange(
      128242n,
      FullTrackName.tryNew('track/namespace', 'trackName'),
      new Location(81n, 81n),
      100n,
      [new Forward(true), new GroupOrderParam(GroupOrder.Ascending)],
    )
  }

  describe('Subscribe', () => {
    it('should roundtrip correctly', () => {
      const subscribe = buildTestSubscribe()
      const serialized = subscribe.serialize()

      const buf = new ByteBuffer()
      buf.putBytes(serialized.toUint8Array())
      const msgType = buf.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Subscribe))

      const msgLength = buf.getU16()
      expect(msgLength).toBe(buf.remaining)

      const deserialized = Subscribe.parsePayload(buf)
      expect(deserialized).toEqual(subscribe)
      expect(buf.remaining).toBe(0)
    })

    it('should roundtrip with excess trailing bytes', () => {
      const subscribe = buildTestSubscribe()
      const serialized = subscribe.serialize()
      const extra = new Uint8Array([...serialized.toUint8Array(), 9, 1, 1])

      const buf = new ByteBuffer()
      buf.putBytes(extra)

      const msgType = buf.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Subscribe))

      const msgLength = buf.getU16()
      expect(msgLength).toBe(buf.remaining - 3)

      const deserialized = Subscribe.parsePayload(buf)
      expect(deserialized).toEqual(subscribe)

      const trailing = buf.toUint8Array().slice(buf.offset)
      expect(Array.from(trailing)).toEqual([9, 1, 1])
    })

    describe('Subscribe Constructors', () => {
      it('should create a Subscribe with AbsoluteRange filter in parameters', () => {
        const subscribe = Subscribe.newAbsoluteRange(
          128242n,
          FullTrackName.tryNew('track/namespace', 'trackName'),
          new Location(81n, 81n),
          100n,
          [],
        )

        const filter = subscribe.parameters.find((p): p is SubscriptionFilter => p instanceof SubscriptionFilter)
        expect(filter?.filterType).toBe(FilterType.AbsoluteRange)
        expect(filter?.startLocation).toEqual(new Location(81n, 81n))
        expect(filter?.endGroup).toBe(100n)
      })

      it('should throw an error if EndGroup < StartGroup', () => {
        expect(() =>
          Subscribe.newAbsoluteRange(
            128242n,
            FullTrackName.tryNew('track/namespace', 'trackName'),
            new Location(81n, 81n),
            80n,
            [],
          ),
        ).toThrow('End Group must be >= Start Group')
      })
    })

    it('should handle empty parameters', () => {
      const subscribe = Subscribe.newLatestObject(128242n, FullTrackName.tryNew('track/namespace', 'trackName'), [])

      const serialized = subscribe.serialize()
      const buf = new ByteBuffer()
      buf.putBytes(serialized.toUint8Array())

      const msgType = buf.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Subscribe))
      buf.getU16()

      const deserialized = Subscribe.parsePayload(buf)
      expect(deserialized).toEqual(subscribe)
      expect(buf.remaining).toBe(0)
    })

    it('should throw on partial message', () => {
      const subscribe = buildTestSubscribe()
      const serialized = subscribe.serialize()
      const serializedBytes = serialized.toUint8Array()

      const partialBytes = serializedBytes.slice(0, Math.floor(serializedBytes.length / 2))
      const buf = new ByteBuffer()
      buf.putBytes(partialBytes)

      try {
        buf.getVI()
        buf.getU16()

        expect(() => Subscribe.parsePayload(buf)).toThrow()
      } catch (err) {
        expect(err).toBeInstanceOf(Error)
      }
    })
  })
}

File: libs/moqtail-rs/src/model/control/subscribe.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::constant::{ControlMessageType, FilterType, GroupOrder};
use super::control_message::ControlMessageTrait;
use crate::model::common::location::Location;
use crate::model::common::tuple::{Tuple, TupleField};
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::data::full_track_name::FullTrackName;
use crate::model::error::ParseError;
use crate::model::parameter::message_parameter::{
  MessageParameter, deserialize_message_parameters,
};
use bytes::{Buf, BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct Subscribe {
  pub request_id: u64,
  pub track_namespace: Tuple,
  pub track_name: TupleField,
  pub subscribe_parameters: Vec<MessageParameter>,
}

impl Subscribe {
  fn build_base_params(
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
  ) -> Vec<MessageParameter> {
    let mut params = vec![
      MessageParameter::new_subscriber_priority(subscriber_priority),
      MessageParameter::new_forward(forward),
    ];
    if !matches!(group_order, GroupOrder::Original) {
      params.push(MessageParameter::new_group_order(group_order));
    }
    params
  }

  pub fn new_next_group_start(
    request_id: u64,
    track_namespace: Tuple,
    track_name: TupleField,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    subscribe_parameters: Vec<MessageParameter>,
  ) -> Self {
    let mut params = Self::build_base_params(subscriber_priority, group_order, forward);
    params.push(MessageParameter::new_subscription_filter(
      FilterType::NextGroupStart,
      None,
      None,
    ));
    params.extend(subscribe_parameters);
    Self {
      request_id,
      track_namespace,
      track_name,
      subscribe_parameters: params,
    }
  }

  pub fn new_latest_object(
    request_id: u64,
    track_namespace: Tuple,
    track_name: TupleField,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    subscribe_parameters: Vec<MessageParameter>,
  ) -> Self {
    let mut params = Self::build_base_params(subscriber_priority, group_order, forward);
    params.push(MessageParameter::new_subscription_filter(
      FilterType::LatestObject,
      None,
      None,
    ));
    params.extend(subscribe_parameters);
    Self {
      request_id,
      track_namespace,
      track_name,
      subscribe_parameters: params,
    }
  }

  pub fn new_absolute_start(
    request_id: u64,
    track_namespace: Tuple,
    track_name: TupleField,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    start_location: Location,
    subscribe_parameters: Vec<MessageParameter>,
  ) -> Self {
    let mut params = Self::build_base_params(subscriber_priority, group_order, forward);
    params.push(MessageParameter::new_subscription_filter(
      FilterType::AbsoluteStart,
      Some(start_location),
      None,
    ));
    params.extend(subscribe_parameters);
    Self {
      request_id,
      track_namespace,
      track_name,
      subscribe_parameters: params,
    }
  }

  pub fn new_absolute_range(
    request_id: u64,
    track_namespace: Tuple,
    track_name: TupleField,
    subscriber_priority: u8,
    group_order: GroupOrder,
    forward: bool,
    start_location: Location,
    end_group: u64,
    subscribe_parameters: Vec<MessageParameter>,
  ) -> Self {
    assert!(
      end_group >= start_location.group,
      "End Group must be >= Start Group"
    );
    let mut params = Self::build_base_params(subscriber_priority, group_order, forward);
    params.push(MessageParameter::new_subscription_filter(
      FilterType::AbsoluteRange,
      Some(start_location),
      Some(end_group),
    ));
    params.extend(subscribe_parameters);
    Self {
      request_id,
      track_namespace,
      track_name,
      subscribe_parameters: params,
    }
  }

  pub fn get_full_track_name(&self) -> FullTrackName {
    FullTrackName {
      namespace: self.track_namespace.clone(),
      name: self.track_name.clone(),
    }
  }

  /// Returns the SubscriptionFilter parameter if present.
  pub fn get_subscription_filter(&self) -> Option<(FilterType, Option<Location>, Option<u64>)> {
    self.subscribe_parameters.iter().find_map(|p| {
      if let MessageParameter::SubscriptionFilter {
        filter_type,
        start_location,
        end_group,
      } = p
      {
        Some((*filter_type, start_location.clone(), *end_group))
      } else {
        None
      }
    })
  }
}
impl ControlMessageTrait for Subscribe {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::Subscribe)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;

    payload.extend_from_slice(&self.track_namespace.serialize()?);
    payload.put_vi(self.track_name.len())?;
    payload.extend_from_slice(self.track_name.as_bytes());

    payload.put_vi(self.subscribe_parameters.len())?;
    for param in &self.subscribe_parameters {
      payload.extend_from_slice(&param.serialize()?);
    }

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "Subscribe::serialize",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);
    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;
    let track_namespace = Tuple::deserialize(payload)?;

    let name_len_u64 = payload.get_vi()?;
    let name_len: usize = name_len_u64
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "Subscribe::parse_payload(track_name_len)",
        from_type: "u64",
        to_type: "usize",
        details: e.to_string(),
      })?;

    if payload.remaining() < name_len {
      return Err(ParseError::NotEnoughBytes {
        context: "Subscribe::parse_payload(track_name)",
        needed: name_len,
        available: payload.remaining(),
      });
    }
    let track_name = TupleField::new(payload.copy_to_bytes(name_len));

    let param_count = payload.get_vi()?;
    let subscribe_parameters =
      deserialize_message_parameters(payload, param_count, ControlMessageType::Subscribe)?;

    Ok(Box::new(Subscribe {
      request_id,
      track_namespace,
      track_name,
      subscribe_parameters,
    }))
  }
  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::Subscribe
  }
}
#[cfg(test)]
mod tests {
  use super::*;
  use crate::model::control::constant::GroupOrder;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let request_id = 128242;
    let track_namespace = Tuple::from_utf8_path("nein/nein/nein");
    let track_name = TupleField::from_utf8("${Name}");
    let start_location = Location {
      group: 81,
      object: 81,
    };
    let subscribe = Subscribe::new_absolute_range(
      request_id,
      track_namespace,
      track_name,
      31,
      GroupOrder::Original,
      false,
      start_location,
      100,
      vec![],
    );

    let mut buf = subscribe.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Subscribe as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = Subscribe::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let request_id = 128242;
    let track_namespace = Tuple::from_utf8_path("nein/nein/nein");
    let track_name = TupleField::from_utf8("${Name}");
    let start_location = Location {
      group: 81,
      object: 81,
    };
    let subscribe = Subscribe::new_absolute_range(
      request_id,
      track_namespace,
      track_name,
      31,
      GroupOrder::Ascending,
      true,
      start_location,
      100,
      vec![],
    );

    let serialized = subscribe.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Subscribe as u64);
    let msg_length = buf.get_u16();

    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = Subscribe::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let request_id = 128242;
    let track_namespace = Tuple::from_utf8_path("nein/nein/nein");
    let track_name = TupleField::from_utf8("${Name}");
    let start_location = Location {
      group: 81,
      object: 81,
    };
    let subscribe = Subscribe::new_absolute_range(
      request_id,
      track_namespace,
      track_name,
      31,
      GroupOrder::Ascending,
      true,
      start_location,
      100,
      vec![],
    );

    let mut buf = subscribe.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Subscribe as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = Subscribe::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
